crates/matrix-sdk-ui/src/timeline/controller/decryption_retry_task.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright 2025 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::BTreeSet, sync::Arc};
+
+use matrix_sdk::deserialized_responses::TimelineEventKind as SdkTimelineEventKind;
+use tokio::sync::{
+    mpsc::{self, Receiver, Sender},
+    RwLock,
+};
+use tracing::{debug, error, field, info, info_span, Instrument as _};
+
+use crate::timeline::{
+    controller::{TimelineSettings, TimelineState},
+    traits::{Decryptor, RoomDataProvider},
+    EncryptedMessage, TimelineItem,
+};
+
+/// Holds a long-running task that is used to retry decryption of items in the
+/// timeline when new information about a session is received.
+///
+/// Creating an instance with [`DecryptionRetryTask::new`] creates the async
+/// task, and a channel that is used to communicate with it.
+///
+/// The underlying async task will stop soon after the [`DecryptionRetryTask`]
+/// is dropped, because it waits for the channel to close, which happens when we
+/// drop the sending side.
+#[derive(Clone, Debug)]
+pub struct DecryptionRetryTask<D: Decryptor> {
+    /// The sending side of the channel that we have open to the long-running
+    /// async task. Every time we want to retry decrypting some events, we
+    /// send a [`DecryptionRetryRequest`] along this channel. Users of this
+    /// struct call [`DecryptionRetryTask::decrypt`] to do this.
+    sender: Sender<DecryptionRetryRequest<D>>,
+}
+
+/// How many concurrent retry requests we will queue before blocking when
+/// attempting to queue another. We don't normally expect more than one or two
+/// will be queued at a time, so blocking should be a rare occurrence.
+const CHANNEL_BUFFER_SIZE: usize = 100;
+
+impl<D: Decryptor> DecryptionRetryTask<D> {
+    pub(crate) fn new<P: RoomDataProvider>(
+        state: Arc<RwLock<TimelineState>>,
+        room_data_provider: P,
+    ) -> Self {
+        // We will send decryption requests down this channel to the long-running task
+        let (sender, receiver) = mpsc::channel(CHANNEL_BUFFER_SIZE);
+
+        // Spawn the long-running task, providing the receiver so we can listen for
+        // decryption requests
+        matrix_sdk::executor::spawn(decryption_task(state, room_data_provider, receiver));
```

Comment:
Hmm, we're still dropping the `JoinHandle` though the task will get killed because the `sender` gets dropped, which in turn means the receiver will get a `None`.

That's a bit implicit compared to holding on to the `JoinHandle` and cancelling in the `drop()` call of `DecryptionRetryTask, but I guess there's nothing wrong with it.

Reply:
Yeah, I tried to join in `drop()` but it's async, so I can't do it.

Reply:
Oh, you wouldn't await the join in the `drop()` call, you would just `abort()` the task. But perhaps it's cleaner to let the task naturally die due to the `Sender` being gone.

Reply:
I think keeping the join handle and manually aborting shows a clear ownership of the task, and would be a tiny bit clearer, even in terms of discoverability: we'd see the `JoinHandle` in the struct, thus know that this is a task's owner.

Reply:
Fair; my argument in terms of discoverability still holds IMO, can we store the task in the struct at least?

Reply:
Sure, that's probably a wise idea.

Reply:
Done in f39a969da3